File: groupingAnagrams.java
// Time Complexity :O(NK) where K is length of a string
// Space Complexity :O(n) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    // Array containing the first 26 prime numbers, each corresponding to a letter in the alphabet
    int[] primes = {
        2, 3, 5, 7, 11, 13, 17, 19, 23, 29,
        31, 37, 41, 43, 47, 53, 59, 61, 67, 71,
        73, 79, 83, 89, 97, 101
    };


    private double stringProduct(String s) {
        double result = 1;
        for (int i = 0; i < s.length(); i++) {
            result *= primes[s.charAt(i) - 'a'];
        }
        return result;
    }


    public List<List<String>> groupAnagrams(String[] strs) {
        // HashMap to store the product of prime numbers as the key and list of anagrams as the value
        HashMap<Double, List<String>> map = new HashMap<>();

        for (String s : strs) {
            // Calculate the product of primes for the current string
            double keyStringProduct = stringProduct(s);
            // If the product is not already a key in the map, add it with an empty list
            if (!map.containsKey(keyStringProduct)) {
                map.put(keyStringProduct, new ArrayList<>());
            }
            // Add the current string to the list corresponding to its product key
            map.get(keyStringProduct).add(s);
        }
        // Return the grouped anagrams as a list of lists
        return new ArrayList<>(map.values());
    }
}

File: GroupingAnagrams.py
# Time Complexity : O(nk log(k)), where n is the number of strings in strs, and k is the avg length of a string in strs
# Space Complexity : O(nk), , where n is the number of strings in strs, and k is the avg length of a string in strs
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : NA

# Your code here along with comments explaining your approach:
# using one hashmap to solve this problem where the key is the sorted string and the value is a list of all original string sthat if sorted give the same sorted string
# first, taking every string from the given list of strings and sorting them, then checking if the key already exists in the map
# if it does not exist, we initialize it as a list of strings and add the original string as the value in the value list
# if it exists, we append the current original string to the value list (because one string is already mapped and the list is already initialized)

class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        anagram = {}

        if (strs is None or len(strs) == 0):
            return anagram 
        
        for i in range(len(strs)):
            # take a string from strs and sort it
            sortedStr = sorted(strs[i]) # returns a list
            # so we need to convert this list back to a string
            sortedStr = ''.join(sorted(strs[i]))
            
            # now we check if the sortedStr is present as a key
            if sortedStr not in anagram:
                # key does not exist
                # so we add it
                anagram[sortedStr] = [strs[i]] # initializing as a list of actual strings
            else:
                # we append the current original string to the value list of the sorted string key
                anagram[sortedStr].append(strs[i])

        # returning the list of all values in the anagram map
        return list(anagram.values())



        
